activerecord/lib/active_record/querying.rb

```diff
@@ -44,8 +44,12 @@ module Querying
     #   Post.find_by_sql ["SELECT body FROM comments WHERE author = :user_id OR approved_by = :user_id", { :user_id => user_id }]
     def find_by_sql(sql, binds = [], preparable: nil, &block)
       result_set = connection.select_all(sanitize_sql(sql), "#{name} Load", binds, preparable: preparable)
-      column_types = result_set.column_types.dup
-      attribute_types.each_key { |k| column_types.delete k }
+      column_types = result_set.column_types
+
+      unless column_types.empty?
+        column_types = column_types.reject { |k, _| attribute_types.key?(k) }
+      end
```

Comment:
How expensive is calling `reject` on an empty hash? It seems like it shouldn't be much slower than calling `empty?`.

If so maybe this could be reduced to one line:

```suggestion
      column_types = result_set.column_types.reject { |k, _| attribute_types.key?(k) }
```

Reply:
I'd like to avoid extra hash allocation too if possible.

Reply:
```ruby
Benchmark.ips do |x|
  x.report("empty guard") do
    column_types = {}
    unless column_types.empty?
      column_types = column_types.reduce { |k, _| k }
    end
  end

  x.report("reduce directly") do
    column_types = {}
    column_types = column_types.reduce { |k, _| k }
  end

  x.compare!
end
```

```
Warming up --------------------------------------
         empty guard   299.532k i/100ms
     reduce directly   205.423k i/100ms
Calculating -------------------------------------
         empty guard     12.203M (± 8.2%) i/s -     60.505M in   5.003606s
     reduce directly      4.038M (± 4.0%) i/s -     20.337M in   5.043675s

Comparison:
         empty guard: 12203137.3 i/s
     reduce directly:  4038411.1 i/s - 3.02x  slower
```

Reply:
👍